src/app/plugins/connectfour.plugin.ts

```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
+  }
+
+  private _dropPiece(col: number, currentPlayer?: number): boolean {
+    let row = this._getFirstObjectInColumn(col);
+    // Place object one row above first object.
     row--;
 
     if (row === -1) {
       return false;
     }
 
-    this._board[row][col] = this.currentPlayer;
+    this._board[row][col] = currentPlayer ? currentPlayer : this.currentPlayer;
 
     return true;
   }
 
-  public checkWin(): boolean {
-    // Returns if a postion is valid and owned by the current player.
-    const validAndOwned: (row: number, col: number) => boolean = (row, col) => {
-      return (
-        row >= 0 &&
-        row < this._rows &&
-        col >= 0 &&
-        col < this._cols &&
-        this._board[row][col] === this.currentPlayer
-      );
-    };
+  private _removeTopPiece(col: number): void {
+    const row = this._getFirstObjectInColumn(col);
+
+    if (row === this._rows) {
+      return;
+    }
+
+    this._board[row][col] = 0;
+  }
+
+  private _getFirstObjectInColumn(col: number): number {
+    // Find first obstruction from top down in column.
+    let row = 0;
+    while (row < this._rows && this._board[row][col] === 0) {
+      row++;
+    }
+    return row;
+  }
+
+  private _longestChainOnBoard(currentPlayer?: number): number {
+    return this._board.reduce((longestChainOnBoard, rowObj, row) => 
+      rowObj.reduce((longestChainStartingInRow, _, col) => 
+      Math.max(longestChainStartingInRow, this._longestChainAtLocation(row, col, currentPlayer))
+      , longestChainOnBoard)
+    , 0);
+  }
 
-    const chainOfFourInAnyDirection: (row: number, col: number) => boolean = (row, col) => {
-      const chains: number[] = Array(8).fill(0);
+  private _longestChainAtLocation(row: number, col: number, currentPlayer?: number): number {
+    const chains: number[] = Array(8).fill(0);
 
-      // For all locations in chain of length four
-      for (let distance = 0; distance < 4; distance++) {
-        // For each possible chain direction
-        for (let direction = 0; direction < 8; direction++) {
-          const newRow: number = row + this._searchDY[direction] * distance;
-          const newCol: number = col + this._searchDX[direction] * distance;
+    // For all locations in chain of length four
+    for (let distance = 0; distance < 4; distance++) {
+      // For each possible chain direction
+      for (let direction = 0; direction < 8; direction++) {
+        const newRow: number = row + this._searchDY[direction] * distance;
+        const newCol: number = col + this._searchDX[direction] * distance;
 
-          if (validAndOwned(newRow, newCol)) {
-            chains[direction]++;
-          }
+        if (this._validAndOwned(newRow, newCol, currentPlayer)) {
+          chains[direction]++;
         }
       }
+    }
 
-      // Return whether or not we found a chain of four.
-      return chains.includes(4);
-    };
+    // Return whether or not we found a chain of four.
+    return Math.max(...chains);
```

Comment:
Do we need the spread operator? I thought `Math.max` could take in a `number[]` as is

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/max